warp/aggregator/aggregation_job.go

```diff
@@ -0,0 +1,139 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package aggregator
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/set"
+	avalancheWarp "github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type signatureAggregationJob struct {
+	// SignatureBackend is assumed to be thread-safe and may be used by multiple signature aggregation jobs concurrently
+	client   SignatureBackend
+	height   uint64
+	subnetID ids.ID
+
+	// Minimum threshold at which to return the resulting aggregate signature. If this threshold is not reached,
+	// return an error instead of aggregating the signatures that were fetched.
+	minValidQuorumNum uint64
+	// Threshold at which to cancel fetching further signatures
+	maxNeededQuorumNum uint64
+	// Denominator to use when checking if we've reached the threshold
+	quorumDen uint64
+	state     validators.State
+	msg       *avalancheWarp.UnsignedMessage
+}
+
+type AggregateSignatureResult struct {
+	SignatureWeight uint64
+	TotalWeight     uint64
+	Message         *avalancheWarp.Message
+}
+
+func NewSignatureAggregationJob(
```

Comment:
Good point, this can be internal as well

Reply:
Changed to internal, but looking at this, it may pass around a few parameters, but passing the aggregator as an input creates a dependency on the aggregator for the signature aggregation job tests that can be avoided otherwise, so going to leave this as is.